[WIP] Feature/game play

serpient opened the pull request:
![game play](https://media.giphy.com/media/RfNNlyoKdi8CHEC604/giphy.gif)

serpient committed:
initialize Board class

serpient committed:
add board methods

serpient committed:
first iteration of player interface and duck typing

- about to change because this was an incorrect implementation of interfaces and duck typing.
- Player classes is trying to be one class keeping track of both player types and player state.

serpient committed:
break apart player and move state to game_state

- 2 instances of player in game state
- player is an interface class that takes in player_type (human or computer) and routes it to the right method.
- game_state also takes in board size and initializes a new board class
- game state keeps track of current player and knows to switch them.

serpient committed:
create game presenter interface and 2 duck types

- ConsoleIO
- TestIO

serpient committed:
set presenter in gameState; adjust move to take in board and presenter

serpient committed:
add validator module; add position input validator

serpient committed:
implement board transformers

serpient committed:
fix board transformer bugs

- needed to start from 0 index.
- generate row values were dynamically creating numbers, but needed to use those as indexes to retrieve the right value from board state

serpient committed:
add game rules module with win and tie checker

serpient committed:
rename board state to positions

serpient committed:
set implementation fns to private

serpient committed:
use Token module for empty/players

serpient committed:
update readme with exe how to; file directory

serpient committed:
add board size and opponent type validators

serpient committed:
create executable to start game

serpient committed:
remove presenter from game state (unnecessary)

serpient committed:
create game setting interface

- opponentType and boardSize are created to hook in.
- should be easy to create other game setting types

serpient committed:
add additional test inputs to testIO

- instead of returning a single value, we are now going to return a series of ordered values so when I ping against it multiple times, it returns an expected different number.
- uses an iterator to keep track

serpient committed:
create messages class with starter banner string

serpient committed:
create starter game class with settings hook

aaizuss commented on lib/board.rb:
```diff
@@ -0,0 +1,41 @@
+
+    def update(position:, token:)
+        positions[position] = token
+        @positions = positions;
```
Do you want this function to update the positions attribute *and* return the updated positions? How do you expect to use this function? I notice that the test just tests that the function returns the updated positions array, which is fine. Line 15 seems redundant. If using the @ symbol is necessary to update the attribute, then you should change line 14 to be `@positions[position] = token`

aaizuss commented on lib/board.rb:
```diff
@@ -0,0 +1,41 @@
+
+    def empty_positions()
+        positions.each_with_index.reduce([]) do |empty_positions, (position, idx)| 
+            position == empty && empty_positions.push(idx)
```
It's neat that you're using reduce here, but I'm not sure that the `&&` is a rubyish way to do things - it reminds me of javascript. If it's common to do this in Ruby, ignore my comment :)

aaizuss replied:
why don't you take advantage of the `is_position_empty` function here? (it's a is a very simple line though so I don't have a strong opinion)

aaizuss commented on lib/game_presenter.rb:
```diff
@@ -0,0 +1,70 @@
+        presenter.get_input(input: input)
+    end
+
+    def output_message(message:)
```
I assume you are using keyword arguments so that the methods are extra clear where they are used, but I'm not sure whether it's helpful when the method takes a single argument and the name of the method is already descriptive.

serpient commented on lib/game.rb:
```diff
@@ -0,0 +1,52 @@
+        setting_results = set_game_settings(settings: settings)
+
+        @game_state = GameState.new(
+            player_1: Human.new(),
```
possibly should be sending in "H" or "C" instead of the actual duck type classes Human or Computer.

serpient commented on lib/validator.rb:
```diff
@@ -0,0 +1,15 @@
+module Validator
+    def position_input_valid?(input:, board:) 
```
could rename to `position_valid?`

serpient commented on lib/game.rb:
```diff
@@ -0,0 +1,52 @@
+        end
+    end
+
+    def print_banner()
```
move this to ConsoleIO?

serpient commented on lib/game.rb:
```diff
@@ -0,0 +1,52 @@
+                game_presenter.output_message(message: setting_type.message())
+                input = game_presenter.get_input() 
+            end
+            map[setting_type.name] = input
```
maybe input should be "cleaned" by setting_type as well?
board size , should convert it to string
opponent type, can do nothing

so it doesnt need to be done in start(), when instantiating the new GameState class

serpient committed:
simplify logic in board update and empty_positions

serpient committed:
add clean input method to game setting classes

- its in charge of "cleaning" the valid input. For example, if board size needs to be a number, it will convert it here.

serpient committed:
move player logic into GameState; remove message keyword

serpient committed:
remove keyword in get_input

- unnecessary as its obvious its input

serpient reviewed (commented):
additional comments

serpient committed:
rename position validation

jayden commented on lib/board.rb:
```diff
@@ -0,0 +1,39 @@
+        @positions[position] = token
+    end
+
+    def is_full?()
```
In idiomatic Ruby, we usuallly omit `is` as a prefix in methods that return a boolean `is_full? -> full?` 

Similarly, we omit the parentheses when a method doesn't take any parameters.

jayden commented on lib/board.rb:
```diff
@@ -0,0 +1,39 @@
+        end
+    end
+
+    def is_input_valid?(int)
```
Does validation belong in the board?

jayden commented on lib/board_transformer.rb:
```diff
@@ -0,0 +1,34 @@
+    def vertical_rows(board)
+        max_positions = board.max_positions
+        size = board.size 
+        (0..size - 1).map do |starting_row|
```
This can be `(0...size)`. Representing a range with two dots will include the last value and omit it if you use three dots.

serpient committed:
extract "H" and "C" to player_type module

jayden commented on lib/game_presenter.rb:
```diff
@@ -0,0 +1,70 @@
+    end
+end
+
+class TestIO
```
Why are we nesting three classes in here? Is this just used for testing purposes?

serpient replied:
I've been keeping duck type classes with the interface it works with in one file. I wasn't sure how its usually organized, so kept it together for easy changes

serpient replied:
But I did just move TestIO into the spec folder, since its pretty much used only on the testing side

serpient committed:
extract TestIO into spec folder

jayden commented on lib/game.rb:
```diff
@@ -0,0 +1,49 @@
+        @game_presenter = game_presenter
+    end
+
+    def set_game_settings(settings:)
```
How else can we *inject* your game settings without this setter method?

serpient replied:
Do you mean .... moving all this logic into a SetGameSettings or GameStarter class, that calls the different Game Settings? then using the results in the `start()` ?

serpient committed:
rename fns to be more ruby like

- .. vs ... in ranges
- In idiomatic Ruby, we usuallly omit is as a prefix in methods that return a boolean is_full? -> full?
- also, omit the parentheses when a method doesn't take any parameters.

serpient committed:
change [0] to first method

- [0] keeps it tied to a array structure;
- first can be used for any ENUM type

serpient commented on lib/game.rb:
```diff
@@ -0,0 +1,47 @@
+            input = nil
+            while !setting_type.valid?(input: input)
+                print_banner
+                game_presenter.output_message(setting_type.message)
```
print_banner and game_presenter output message, should all be bundled into something like `output_game_settings` and placed into GamePresenter??

serpient commented on lib/game.rb:
```diff
@@ -0,0 +1,47 @@
+    end
+
+    def start()
+        settings = [
```
maybe on Tic Tac Toe execution, 
run a class like GameSettingPicker, which takes care of grabbing game setting input.

then we can pass `self` of GameSettingPicker into `Game` ??
the results of the game setting are then used to initialize game_state in `Game`, 
then `play` runs the game loop.

serpient commented on lib/board.rb:
```diff
@@ -0,0 +1,35 @@
+        positions.all? {|position| position != empty }
+    end
+
+    def position_is_available?(position)
```
rename to `position_available?`

aaizuss replied:
Yes

serpient commented on lib/game.rb:
```diff
@@ -0,0 +1,47 @@
+
+        @game_state = GameState.new(
+            player_2: setting_results["opponent_type"],
+            board_size: setting_results["board_size"]
```
a little weird that we know the exact "keys" needed to grab the right result in setting_results.

serpient commented on lib/game_state.rb:
```diff
@@ -0,0 +1,28 @@
+    )
+        @board = Board.new(size: board_size)
+        @player_1 = Player.new(player: get_player(player_1).new(), token: hero)
+        @player_2 = Player.new(player: get_player(player_2).new(), token: opponent)
```
is it weird that Game state knows how `Board` and `Player` should be initialized ? 
is this too dependent?

aaizuss replied:
I think it's fine

serpient committed:
extract game setting IO into game presenter;

- also rename GameSettings to GameSetting

serpient committed:
create Game Setting Setter classl hook to Game

serpient committed:
update game; add turn function

- fix minor bugs and imports
- add update fn to gameState
- add choose position message
- fix board range validation

aaizuss commented on README.md:
```diff
@@ -1,5 +1,22 @@
+# Setting up executable
+[Cli Application Reading](https://learn.co/lessons/ruby-cli-applications-readme)
+
+1. Place any starting executable files in /bin folder
```
Why are these steps in your readme? Your readme should include steps for a user to run the program, which is just step 3.

aaizuss commented on lib/board_transformer.rb:
```diff
@@ -0,0 +1,42 @@
+        return board.positions.each_slice(board.size).to_a
+    end
+    
+    def vertical_rows(board)
```
there's a simpler way to get columns... hint: look into `transpose`

aaizuss replied:
I think rows and columns are clearer names for horizontal_rows and vertical_rows - what do you think?

serpient replied:
Ok, changing!

aaizuss commented on lib/game.rb:
```diff
@@ -0,0 +1,48 @@
+    end
+
+    def play
+        4.times { turn() }
```
why 4 times?

serpient replied:
this function is still a WIP -> i was testing TestIO amongst a few other things.

aaizuss commented on lib/game_rules.rb:
```diff
@@ -0,0 +1,30 @@
+
+        diagonal_wins = any_row_a_win?(diagonal_rows(board))
+
+        return [ horizontal_wins, vertical_wins, diagonal_wins ].any?
```
Instead of creating an array and calling a function, you can simplify this by taking advantage of the or operator: `horizontal_wins || vertical_wins || diagonal_wins`

aaizuss commented on lib/game_rules.rb:
```diff
@@ -0,0 +1,30 @@
+    end
+
+    def tie?(board:)
+        return board.full?
```
A full board does not necessarily mean there is a tie...

serpient replied:
wait - why not? I would check for tie after wins -- so if its not a win, and the board is full -- wouldnt that be a tie?

jayden replied:
I think that's the point though - we aren't checking for a win inside `tie?` and you're leaving it up to any client of `GameRules` to assume that they need to check for `win?` first.

serpient replied:
ahh i see, ok , updating!

aaizuss commented on bin/tic_tac_toe.rb:
```diff
@@ -0,0 +1,13 @@
+
+game = Game.new(
+    board_size: game_settings.settings["board_size"],
+    player_2: game_settings.settings["opponent_type"],
```
`"board_size"` and `"opponent_type"` seem like a magic strings

aaizuss commented on lib/validator.rb:
```diff
@@ -0,0 +1,23 @@
+        int >= 0 && int < board.max_positions
+    end
+    
+    def board_size_valid?(input:)
```
Maybe this function should use info from `Board` to determine whether the board size is valid

aaizuss commented on lib/token.rb:
```diff
@@ -0,0 +1,13 @@
+module Token
+    def hero
+        return :hero
```
What does hero mean?

serpient replied:
It's just the token I'm using for each player. if its not clear, I can go back and do "player_1" and "player_2" ?

jayden replied:
It may be clearer if it's just `X` and `O` as these are the default Tic Tac Toe game markers

aaizuss commented on spec/board_transformer_spec.rb:
```diff
@@ -0,0 +1,82 @@
+
+        it 'converts 3x3 board data to vertical rows' do
+            board = Board.new(size: 3)
+            board.update(position: 0, token: 1)
```
Is there a way to set up the boards for these tests without calling the update function a bunch of times? It would make the tests clearer

serpient committed:
update readme to only have executable instructions

serpient committed:
extract setting types into own module

serpient committed:
update board & board transformer per feedback

- change name from position_is_available? to position_available?
- use transpose to create vertical rows.
- rename transformer to just horizontal, vertical and diagonal
- change how diagonals are being created
- add new arg in board update, so we can pass in an entire array

jayden commented on lib/player.rb:
```diff
@@ -0,0 +1,34 @@
+class Player
```
Ideally, we should separate classes into their own separate files as it scales better and is easier for others to navigate the codebase. If this project were to get more complex, files like this can grow quite large.

Instead of a flat project structure, we can introduce a `players` directory and create a `Players` module.

```ruby
# lib/players/player.rb
module Players
  class Player
  end
end

# lib/players/human.rb
module Players
  class Human
  end
end

# lib/players/computer.rb
module Players
  class Computer
  end
end
```

jayden commented on lib/messages.rb:
```diff
@@ -0,0 +1,21 @@
+        empty = "                                     \n                                     \n"
+        title = "         T I C  T A C  T O E         \n"
+
+        return "\n#{border}#{empty}#{title}#{empty}#{border}\n"
```
Bit nitpicky, but remember that Ruby returns the last value in a method so declaring `return` here is unnecessary.

jayden commented on lib/game_setting_setter.rb:
```diff
@@ -0,0 +1,33 @@
+    )
+        @setting_types = setting_types
+        @game_presenter = game_presenter
+        @settings = gets
```
It's a bit strange that we are calling `gets` inside this constructor as it has side effects. We should only be initializing a class inside a constructor and leave any actual operations as methods.

cmvandrevala replied:
Agreed

serpient committed:
change tokens to player_x and player_o

serpient committed:
update tie to check for NOT win

- so its not dependent on function call order. before, it relied on win? being checked first before running tie. This way, it checks that its NOT a win and board is full before ascerting a tie.

jayden commented on lib/setting_types.rb:
```diff
@@ -0,0 +1,9 @@
+module SettingTypes
+    def board_size
```
Methods like this that just return a static value can just be declared as constants

```ruby
module SettingTypes
  BOARD_SIZE = :board_size
  OPPONENT_TYPE = :opponent_type
end
```

serpient committed:
add win and tie statuses in turn

serpient committed:
game refactor

- extract hard code strings and put into messages
- clean up move fn - refers to itself to get token and put into message
- move output messaging into player move fn()
- fix incorrect player set as winner bug by putting game status check into if/else block.

serpient committed:
move around project structure

interfaces are folders with associated duck types

serpient committed:
remove all unecessary returns

- Ruby returns the last value in a method so declaring return is unnecessary.

serpient committed:
extract update game status fn

serpient committed:
move and update game status  are set as private

cmvandrevala commented on README.md:
```diff
@@ -1,5 +1,15 @@
 # Run tests
-`rspec`
+`guard`
```
I would have both the `rspec` and `guard` commands here, just in case you want to use a test runner or just a single-run.

cmvandrevala commented on README.md:
```diff
@@ -1,5 +1,15 @@
+`bundle install`
+
+# Running the executable
+`ruby bin/tic_tac_toe.rb`
```
If `bin/tic_tac_toe.rb` contains a shebang (i.e. `#!/usr/bin/env ruby`), can't you just run `./bin/tic_tac_toe.rb`? No need for the ruby command beforehand?

cmvandrevala commented on bin/tic_tac_toe.rb:
```diff
@@ -0,0 +1,16 @@
+
+game = Game.new(
+    board_size: game_settings.settings[board_size],
+    player_2: game_settings.settings[opponent_type],
```
It is a bit strange to see player two with no player one here...

serpient replied:
I used to have player_1 as well, but it always defaults to Human. So I removed and only ask for board_size and player_2.

I was playing around with names like "Hero" and "Opponent" instead of Player 1 and Player 2 -- would that be more clear?

cmvandrevala replied:
Maybe? I was just thinking if you wanted a computer vs computer option, it might break down.

cmvandrevala commented on lib/board.rb:
```diff
@@ -0,0 +1,35 @@
+require_relative './token'
+
+class Board
+    include Token
```
Be careful. Token has a meaning in software (e.g. authentication token, JSON web token, etc). Maybe use a different term here?

cmvandrevala commented on lib/board.rb:
```diff
@@ -0,0 +1,35 @@
+    def initialize(size:)
+        @size = size
+        @max_positions = size * size
+        @positions = (1..max_positions).map {|num| num = empty }
```
Is `empty` a variable? A string?

serpient replied:
empty comes from the Token module, representing `:empty`. 

I'll be updating these "tokens" (i'll change the name!) to be constants instead, so when reading it as `num = EMPTY` be clearer?

cmvandrevala replied:
I would almost prefer to see something like `Token.EMPTY` so I know it refers to the Token module

jayden replied:
^ I'd add that this goes for any module that you are including. Prepending the module name will make it easier for others to identify where things outside of the current scope are coming from.

serpient replied:
I'm having trouble prepending module definitions. For example, when I write
```
include GameSettings

GameSettings.board_size 
```
this gives me a `NoMethodError`. How do I prepend with methods??

esmevane replied:
I think instead of this:

```
module GameSettings
    def board_size
        {}
    end
end 

include GameSettings

GameSettings.board_size 
```

...you might be able to do this:

```
module GameSettings
  extend self
  
  def board_size
    {}
  end
end
```

Because everything in Ruby is an object, `GameSettings` can actually add those functions to its own topmost level.  Under the hood this is kind of like the code is saying `GameSettings = Module.new(GameSettings)`, and then `GameSettings.include(GameSettings)`.  That's not _exactly_ what's happening but it is pretty similar.

The end result should be that you can do stuff like this:

`@size = GameSettings.board_size`

cmvandrevala commented on lib/board.rb:
```diff
@@ -0,0 +1,35 @@
+        @positions = (1..max_positions).map {|num| num = empty }
+    end
+
+    def update(position:, token:, all_positions: nil)
```
Why pass in `all_positions` when you already have `@positions`?

serpient replied:
all_positions was something I passed in for easy testing.

so instead of updating one position at a time, I can pass in a new array during tests.

Should I just make a new function just for tests?

jayden replied:
We should not create new versions of functions in your implementation for easy testability. You could consider creating a helper method in your tests instead.

```ruby
def fill_board(board, token, positions)
  positions.each { | position| board.update(position, token) }
end
```

cmvandrevala commented on lib/board_transformer.rb:
```diff
@@ -0,0 +1,28 @@
+    end
+    
+    def vertical(board)
+        horizontal(board).transpose()
```
Nice. I like the use of transpose here

cmvandrevala commented on lib/board_transformer.rb:
```diff
@@ -0,0 +1,28 @@
+        horizontal(board).transpose()
+    end
+
+    def diagonal(board)
```
Are you dealing with boards of any size or just 3x3? If you are just dealing with 3x3 boards, why not just hard-code the diagonals?

serpient replied:
i'm dealing with 3x3 and 4x4 boards. Should I try hard coding for the 2 board sizes? Or keep the diagonal generator?

cmvandrevala commented on lib/game.rb:
```diff
@@ -0,0 +1,55 @@
+
+    def play
+        while status == :play
+            turn
```
Maybe rename this to `take_turn`? Turn makes me think of a turning wheel

cmvandrevala commented on lib/game.rb:
```diff
@@ -0,0 +1,55 @@
+        input = move while !position_valid?(input: input, board: game_state.board)
+        game_state.update(position: input.to_i)
+        update_game_status
+        game_state.switch_players if status == :play
```
I think you can refactor this to make it much more expressive. Maybe:

```
def turn
  get_valid_move()
  update_game_status()
  if game_is_over()
    show_the_winner()
  else
    switch_players()
  end
end
```

cmvandrevala commented on lib/game.rb:
```diff
@@ -0,0 +1,55 @@
+    end
+
+    private
+    def move
```
I would consider adding a linter like rubocop to the project to help with code formatting

cmvandrevala commented on lib/game_presenter/console_io.rb:
```diff
@@ -0,0 +1,24 @@
+require_relative '../messages'
+
+class ConsoleIO
+    include Messages
```
I don't think you should `include Messages` here. Instead, have the `ConsoleIO` class literally just be a few methods that are in charge of printing strings.

cmvandrevala commented on lib/game_presenter/game_presenter.rb:
```diff
@@ -0,0 +1,25 @@
+        @presenter = presenter
+    end
+
+    def get_input(input=nil)
```
Why does the presenter have a `get_input` method? I think that would be the responsibility of the `ConsoleIO` class

serpient replied:
So game-presenter was meant to be an interface that has duck types of ConsoleIO and TestIO (am i using the right terms?). 
So is the naming  confusing but the implementation ok? Or I should still separate these

cmvandrevala commented on lib/game_presenter/test_io.rb:
```diff
@@ -0,0 +1,44 @@
+    attr_accessor :count
+
+    def initialize()
+        @count = 0
```
Did you create a spy mock object here? That is so cool!!! Nice job!!!

cmvandrevala commented on lib/game_rules.rb:
```diff
@@ -0,0 +1,27 @@
+        horizontal_wins = any_row_a_win?(horizontal(board))
+        vertical_wins = any_row_a_win?(vertical(board))
+        diagonal_wins = any_row_a_win?(diagonal(board))
+        horizontal_wins || vertical_wins || diagonal_wins
```
`any_row_a_win?(horizontal(board)) || any_row_a_win?(vertical(board)) || any_row_a_win?(diagonal(board))`

esmevane replied:
Personally I dig the clarity here, just my two cents though. 💰

cmvandrevala commented on lib/game_setting/game_setting.rb:
```diff
@@ -0,0 +1,28 @@
+    end
+
+    def valid?(input: nil)
+        @input = input
```
If you never return `@input` and you never use it within the class, why have it? Can't this just be `setting.valid?(input: input)`?

serpient replied:
you're right! I think I was trying something and forgot to remove this

cmvandrevala commented on lib/game_setting/opponent_type.rb:
```diff
@@ -0,0 +1,23 @@
+    end
+
+    def clean(input:)
+        input
```
Why have this method? What is it cleaning?

serpient replied:
So clean() is meant to convert the input to whatever the rest of the application expects.
For example, for board size input, it converts it to a integer and passes it back.

Opponent type doesn't do anything, but I didnt know how else to do it if i wanted swappable duck types. I'm not sure if I'm implementing duck types correctly for my game_setting types

cmvandrevala commented on lib/game_setting/setting_types.rb:
```diff
@@ -0,0 +1,9 @@
+module SettingTypes
```
Why do you have the `SettingTypes` module?

serpient replied:
I have a few places in my app that relies on the string types. So I stored it in one place instead of hard coding strings in 2 places (tic_tac_toe)

cmvandrevala commented on lib/messages.rb:
```diff
@@ -0,0 +1,33 @@
+        empty = "                                     \n                                     \n"
+        title = "         T I C  T A C  T O E         \n"
+
+        "\n#{border}#{empty}#{title}#{empty}#{border}\n"
```
I would put `border`, `empty`, and `title` into their own methods

cmvandrevala commented on lib/player/player_types.rb:
```diff
@@ -0,0 +1,9 @@
+module PlayerType
+    def human
+        "H"
```
Why use the string labels? Why not just have a `HumanPlayer` class and a `ComputerPlayer` class?

serpient replied:
this is the valid inputs for when setting a player type before game play. 
I use "H" and "C" a few places in the code , including tests, so I kept them here.  is that ok?

cmvandrevala commented on lib/token.rb:
```diff
@@ -0,0 +1,13 @@
+module Token
+    def player_x
+        :X
```
No need to change this right now, but what if I wanted to be able to choose my symbol on the TTT board?

serpient replied:
This part wouldn't change, as it's not tied to a string / rendered board -- this is just for our internal data.  I would create a new setting type class, get the input and store in game state. then refer to it when I create a board output. 

Would changing the name to... maybe Symbols or something help?

esmevane replied:
I do like Symbols.  There's a good amount of valid names here, probably, and Token/Tokens is definitely one.  Naming is rough.

cmvandrevala commented on spec/board_spec.rb:
```diff
@@ -0,0 +1,84 @@
+require_relative '../lib/token'
+
+describe 'Board' do
+    include Token
```
Why do you need to `include Token` in these tests?

serpient replied:
I included Token so I dont have hard coded symbols/strings throughout my code. It'll always match the tokens that I use in the production code.
is that ok? or are you saying that the use of Tokens are ok , but not needed in this specific set of tests, at which point, I did want it to check a few functions that ouput the tokens specifically (like new board)

esmevane replied:
This is my guess, but I'd say the concept of `include`ing or `extend`ing can raise some flags for folks because they change the method / instance surface area of the object you're working on.  So, basically, if the other module changes at some point, you have to make sure there's no collisions or unpredictable dependencies.  Good test discipline makes that not as much of a problem, but it can still come up and bite you at random times.

The `extend self` pattern I describe above helps with that because it calls out where every value is coming from, which at least lowers the barrier for a future reader who might not know where everything is mapped to/from.

cmvandrevala commented on spec/board_spec.rb:
```diff
@@ -0,0 +1,84 @@
+
+        it 'returns true when board is full' do
+            board = Board.new(
+                size: 3
```
Do you have any tests for a 4x4 board? I thought that your board size valid function would allow a board size of 4?

serpient replied:
I do not have any 4x4 board tests -- i'll add them now!

cmvandrevala commented on spec/board_transformer_spec.rb:
```diff
@@ -0,0 +1,83 @@
+require_relative '../lib/board_transformer'
+
+describe 'Board Transformer' do
```
I am not sure if `BoardTransformer` needs to be its own class. Maybe just include these functions in `Board`?

cmvandrevala commented on spec/game_presenter_spec.rb:
```diff
@@ -0,0 +1,41 @@
+        game_presenter = GamePresenter.new(
+            presenter: TestIO.new()
+        )
+        it 'gets input through presenter' do
```
I think the presenter should only be in charge of presenting the board, not getting input.

serpient replied:
Is it a problem of the name? Should I separate input output, or just change the name to GameIO?

cmvandrevala replied:
It might just be the name

cmvandrevala commented on spec/game_rules_spec.rb:
```diff
@@ -0,0 +1,93 @@
+    include GameRules
+    include Token
+
+    context 'win? 2x2' do
```
Why test the 2x2 boards if you have:

```
def board_size_valid?(input:)
        int = input.to_i
        int == 3 || int == 4
end
```

serpient replied:
you're right! unnecessary and should be testing 4x4 instead.

cmvandrevala commented on spec/game_setting_setter_spec.rb:
```diff
@@ -0,0 +1,47 @@
+        GameSetting.new(setting: OpponentType.new()),
+    ]
+
+    before(:all) do
```
I would try to use `before(:each)` as much as possible instead of `before(:all)`. It helps keep your tests isolated

cmvandrevala commented on spec/messages_spec.rb:
```diff
@@ -0,0 +1,7 @@
+require_relative '../lib/messages'
+
+describe 'Messages' do
```
You can delete this file if there are no tests

cmvandrevala commented on spec/token_spec.rb:
```diff
@@ -0,0 +1,18 @@
+describe 'Token' do
+    include Token
+    context 'gets tokens' do
+        it '- player 1' do
```
The descriptions are a bit weird here. I know this reads nice when you include "context" and "it" together, but the description for "it" seems incomplete

serpient committed:
use setting type constants; add notes for refactor

serpient committed:
refactor per feedback

- add rspec and update executable instructions in readme
- rename clean() to parse()
- rename turn to take_turn
- extract functions in take_turn, so its more readable
- rename game_presenter to game_IO to be more descriptive of what it is
- change before tests to :each instead of :all, to keep tests isolated
- replace 2x2 board tests with 4x4 board tests

serpient committed:
refactor game setting

- previously used a duck type / interface for this. But after more thought, its not necessary as the methods arent vastly different from each other, I was just using it as a way to store data in the same format.
- Changed to an object / hash instead that gets iterated through.

serpient committed:
game setting setter doesnt set settings in initialization

- unstable to call gets() in initialization. rearrange code to make it a public method thats called after initialization.
- rename token test units to not use "-"
- rename get to create_settings

serpient committed:
create board presenter and stringbBoard

serpient committed:
integrate board presenter into game

- update string printing to start from 0, not 1
- change board presenter to not initalize with a board. board is passed in to generate function.

serpient committed:
extract fns from game

- extract move and valid_move functions into player
- extract output board logic into game_io

serpient committed:
extract fns from game

- extract move and valid_move functions into player
- extract output board logic into game_io
- move becomes a implementation detail (private). get_valid_move is the new public method. Adjust test cases to reflect

serpient committed:
fix bug where board wasnt updating

- forgot to assign result of get_valid_move to @input to be used everywhere else. Without this, it was just overwriting position 0 over and over again

jayden commented on bin/tic_tac_toe.rb:
```diff
@@ -0,0 +1,17 @@
+include SettingTypes
+
+game_settings = GameSettingSetter.new()
+game_settings.create_settings
```
Something we could consider here is to return `settings` from `create_settings` and retrieve the specific settings from this returned value. It makes the code much easier to read and reason about when return values from methods instead of calling functions that have side effects and reaching into the instance variable to get the settings.

serpient committed:
update create settings to return settings
- it was previously setting the results into a attribute within the class. Now, we will return the new settings hash and use it directly.

esmevane commented on lib/game.rb:
```diff
@@ -0,0 +1,69 @@
+
+    attr_accessor :game_state, :game_io, :status, :input, :board_presenter
+
+    def initialize(game_io: ConsoleIO.new, board_size:, player_2:, board_presenter: StringBoard.new())
```
I notice there are some spots where you're using `()` for functions with no arguments, and other spots where you aren't, is that deliberate?

serpient replied:
Not deliberate! Forgot to go through and check for it -- fixed in new PR

esmevane reviewed (approved):
This is overall some really great stuff, you can write almost any language in Ruby and so it's nice to see some really readable, direct, conventional Ruby.  There's some really complicated things in parts and you've made those portions clean and readable, so really great job there, too.

One thing I want to call your attention to is the use of `require_relative`.  There's not _really_ a way to get around using the file system for require statements in Ruby, but there are some scenarios where using `require_relative` would make changing file locations more expensive in terms of time than other approaches.  I don't think anything needs to change, really, right now, so my suggestion here is to try to explore other `require` / packaging patterns in Ruby if you see an opportunity.
